### PR TITLE
BDRE-02 (BDRE-167, BDRE-198) Hyperlinks added for columns in the DOT

### PR DESCRIPTION
**Pull request by raiashu92**
Links have been added to each column so that on clicking it will further generate the DOT's and add them to the already existing one without refreshing it.

### CODE REVIEW

**Comment by raiashu92** [comment]
@bdrekapil @mishisinku please check and merge.

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Relation.java [hunk 1]
```diff
@@ -152,8 +152,12 @@ public static void updateRelations(List<Table> inTableNodes, List<Table> outTabl
 						Function function = (Function) source;
 						if (function.getOutAlias() != null)
 							outColumnName = function.getOutAlias();
-						else
+						else {
+							if(function.getInNodes().isEmpty()) {
+								//System.out.println("function = " + function);
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Relation.java:157]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 2]
```diff
@@ -109,13 +109,20 @@ public String toDotString() {
 //				"label = \"" + getLabel() + "\"\n" +
 //				"shape = \"" + getShape() + "\"\n" +
 //				"];";
+
+		// struct1 [label=<
+		//<TABLE BORDER="0" CELLBORDER="1" CELLSPACING="0">
+		//<TR><TD>left</TD><TD PORT="f1">mid dle</TD><TD PORT="f2">right</TD></TR>
+		//</TABLE>>];
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:116]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 3]
```diff
@@ -125,11 +132,12 @@ public String toDotString() {
 //				middle += "<f" + (i+1) + ">" + column.getColumnName();
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:132]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 4]
```diff
@@ -125,11 +132,12 @@ public String toDotString() {
...
 //		}
 		for (Column column : getColumns()) {
-			middle.append(column.toDotString() + " | ");
+			middle.append("<TR><TD HREF=\"javascript:getTableName('"+ tableName + "','" + column.getColumnName() +"');\"" + " PORT=\""+column.getColumnName()+ "\" >" + column.getColumnName()  + "</TD></TR> " );
 		}
-		middle = new StringBuilder(middle.substring(0, middle.length() - 3));
+		//middle = new StringBuilder(middle.substring(0, middle.length() - 3));
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:137]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 5]
```diff
@@ -109,13 +109,20 @@ public String toDotString() {
...
 		String tableFullName = getDataBase() + "." + getTableName();
 		LOGGER.info("Table " + getTableName() + " : getColumns().size = " + getColumns().size());
 
-		StringBuilder first = new StringBuilder("\n\"" + tableFullName + "\" [\n" + "label = ");
-
+		StringBuilder first = new StringBuilder("\n\"" + tableFullName + "\"  [\n" + "label =< <TABLE BORDER=\"0\" CELLBORDER=\"1\" CELLSPACING=\"0\"> ");
 		String tableHeader = "Table " + tableFullName;
-		StringBuilder middle = new StringBuilder("\"<" + tableFullName + ">" + tableHeader + " | ");
+		first= first.append("<TR><TD>"+ tableHeader +"</TD></TR>");
+		//StringBuilder middle = new StringBuilder("\"<" + tableFullName + ">" + tableHeader + " | ");
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:124]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 6]
```diff
@@ -109,13 +109,20 @@ public String toDotString() {
...
+		StringBuilder middle=new StringBuilder("");
 //		for (int i = 0; i < getColumns().size(); i++) {
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:126]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 7]
```diff
@@ -43,7 +43,7 @@
 
 	public Table(String tableName, String dataBase, String alias, EntityType entityType, boolean isDatabaseTable) throws ETLException {
 		// eliminate "." from table name
-		super((tableName.split("\\."))[tableName.split("\\.").length-1], "Table", "record", tableName);
+		super((tableName.split("\\."))[tableName.split("\\.").length-1], "Table", "records", tableName);
 //		++counter;
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:47]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java [hunk 8]
```diff
@@ -109,13 +109,20 @@ public String toDotString() {
 //				"label = \"" + getLabel() + "\"\n" +
 //				"shape = \"" + getShape() + "\"\n" +
 //				"];";
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Table.java:111]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java [hunk 9]
```diff
@@ -59,7 +64,9 @@ public Function(String label, List<String> inColumnNameList, String outAlias) {
 	public void setOutNode(Node outNode) {this.outNode = outNode;}
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java:64]: MINOR Move this closing curly brace to the next line. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java [hunk 10]
```diff
@@ -59,7 +64,9 @@ public Function(String label, List<String> inColumnNameList, String outAlias) {
...
 	public List<String> getInColumnNameList() {return inColumnNameList;}
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java:65]: MINOR Move this closing curly brace to the next line. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java [hunk 11]
```diff
@@ -59,7 +64,9 @@ public Function(String label, List<String> inColumnNameList, String outAlias) {
...
 	public void setInColumnNameList(List<String> inColumnNameList) {this.inColumnNameList = inColumnNameList;}
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java:66]: MINOR Move this closing curly brace to the next line. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java [hunk 12]
```diff
@@ -59,7 +64,9 @@ public Function(String label, List<String> inColumnNameList, String outAlias) {
...
-	public String getOutAlias() {return outAlias;}
+	public String getOutAlias() {
+		return outAlias;
+	}
 	public void setOutAlias(String outAlias) {this.outAlias = outAlias;}
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java:70]: MINOR Move this closing curly brace to the next line. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ARightCurlyBraceStartLineCheck)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java [hunk 13]
```diff
@@ -59,7 +64,9 @@ public Function(String label, List<String> inColumnNameList, String outAlias) {
...
 
 	public static void updateFunctions(List<Table> inTableNodes, List<Table> outTableNodes, List<Column> inColumnNodes,
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java:72]: MINOR The Cyclomatic Complexity of this method "updateFunctions" is 14 which is greater than 10 authorized. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
MAJOR Remove the unused method parameter(s) "inTableNodes,outTableNodes,constants,relations". [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS1172)

File: data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java [hunk 14]
```diff
@@ -47,6 +47,11 @@ public Function(String label, List<String> inColumnNameList, String outAlias) {
 //		++counter;
```
Comment (bdre-bda) [inline data-lineage/src/main/java/com/wipro/ats/bdre/lineage/entiity/Function.java:47]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: md-rest-api/src/main/java/com/wipro/ats/bdre/md/rest/TableColumnLineageAPI.java [hunk 15]
```diff
@@ -35,15 +40,35 @@
...
      * It also generates the dot string for visualisation of the links.
      *
-     * @param tableName, colName
-     * @return restWrapper It contains an instance of LineageInfo.
+     * @param lineageTabColInfo
+     * @return restWrapper It contains an instance of LineageTabColInfo.
      */
     @RequestMapping(value = {"", "/"}, method = RequestMethod.GET)
     public
     @ResponseBody
```
Comment (bdre-bda) [inline md-rest-api/src/main/java/com/wipro/ats/bdre/md/rest/TableColumnLineageAPI.java:48]: MINOR Reorder the modifiers to comply with the Java Language Specification. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AModifiersOrderCheck)

File: md-rest-api/src/main/java/com/wipro/ats/bdre/md/rest/TableColumnLineageAPI.java [hunk 16]
```diff
@@ -51,12 +76,11 @@ RestWrapper getByTableCol(@RequestParam(value = "tableName", defaultValue = "nul
...
                 GetDotForTable getDotForTable = new GetDotForTable();
                 dot = getDotForTable.dotGeneratorWithCol(args);
-                LineageTabColInfo lineageInfo = new LineageTabColInfo();
                 LOGGER.debug(dot);
-                lineageInfo.setDot(dot.toString());
-                lineageInfo.setTableName(tableName);
-                lineageInfo.setColName(colName);
-                restWrapper = new RestWrapper(lineageInfo, RestWrapper.OK);
+                lineageTabColInfo.setDot(dot.toString());
+//                lineageTabColInfo.setTableName(tableName);
```
Comment (bdre-bda) [inline md-rest-api/src/main/java/com/wipro/ats/bdre/md/rest/TableColumnLineageAPI.java:81]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: md-rest-api/src/main/java/com/wipro/ats/bdre/md/rest/TableColumnLineageAPI.java [hunk 17]
```diff
@@ -64,18 +88,13 @@ RestWrapper getByTableCol(@RequestParam(value = "tableName", defaultValue = "nul
                 String dot = new String();
                 GetDotForTable getDotForTable = new GetDotForTable();
                 dot = getDotForTable.dotGeneratorWithTable(args);
-                LineageTabColInfo lineageInfo = new LineageTabColInfo();
                 LOGGER.debug(dot);
-                lineageInfo.setDot(dot.toString());
-                lineageInfo.setTableName(tableName);
-                restWrapper = new RestWrapper(lineageInfo, RestWrapper.OK);
+                lineageTabColInfo.setDot(dot.toString());
+//                lineageTabColInfo.setTableName(tableName);
```
Comment (bdre-bda) [inline md-rest-api/src/main/java/com/wipro/ats/bdre/md/rest/TableColumnLineageAPI.java:93]: MAJOR This block of commented-out lines of code should be removed. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ACommentedOutCodeLine)

File: workflow-generator/src/main/java/com/wipro/ats/bdre/wgen/ActionNode.java [hunk 18]
```diff
@@ -179,6 +179,8 @@ public void setProcessInfo(ProcessInfo processInfo) {
         } else if (processInfo.getProcessTypeId() == SPARK_ACTION) {
             SparkActionNode sparkActionNode = new SparkActionNode(this);
             containingNodes.add(sparkActionNode);
+        } else if (processInfo.getProcessTypeId() == CRAWLER_PARENT_ACTION) {
```
Comment (bdre-bda) [inline workflow-generator/src/main/java/com/wipro/ats/bdre/wgen/ActionNode.java:182]: MAJOR Either remove or fill this block of code. [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS00108)

**Comment by bdre-bda** [comment]
SonarQube analysis reported 156 issues:
- MAJOR 89 major
- MINOR 65 minor
- INFO 2 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
- MAJOR [Refactor this method to throw at most one checked exception instead of: org.apache.hadoop.hive.ql.parse.ParseException, org.apache.hadoop.hive.ql.parse.SemanticException](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L68) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS1160)
- MINOR [The Cyclomatic Complexity of this method "handleTokTabOrColOrElse" is 14 which is greater than 10 authorized.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L231) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
- MAJOR [Add a default case to this switch.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L131) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
- MAJOR [Introduce a new variable instead of reusing the parameter "functionString".](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L374) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS1226)
- MINOR [Replace all tab characters in this file by sequences of white-spaces.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS00105)
- MINOR [Refactor this code to not nest more than 3 if/for/while/switch/try statements.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L153) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS134)
- MINOR [Rename this local variable name to match the regular expression '^[a-z][a-zA-Z0-9]*$'.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L237) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS00117)
- MINOR [At most one statement is allowed per line, but 2 statements were found on this line.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L244) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS00122)
- MINOR [Refactor this code to not nest more than 3 if/for/while/switch/try statements.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L157) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS134)
- MAJOR [Move the "<EOF>" string literal on the left side of this string comparison.](https://github.com/WiproOpenSourcePractice/openbdre/blob/e2139c6411dadf995c2feae0ad7d162feb035fea/data-lineage/src/main/java/com/wipro/ats/bdre/lineage/LineageMain.java#L152) [rule](https://sonar.openbdre.org/coding_rules#rule_key=squid%3AS1132)
- ... 127 more

**Pull request merged**